harmony-one/harmony
# Testnet shard reduction

## Issue
Reduce the number of Shards in the test network to 2.

It requires a **hardfork**.

To reduce shards from 4 to 2, it will:
1. In the last block before the feature enable, mark external validator as inactive if they has blskey on shard2 and shard3
2. remove harmony inner blskey on shard2 and shard3 from hardcoded list (via hardfork).

## Test in localnet with external validators:
1. apply this [patch]
2. start localnet: `./test/debug.sh`
3. create external validators via [createvalidator]: `./createValidator.sh`
4. after epoch 4(reduction hardfork in localnet), check if s0/s1 is working properly, and s2/s3 should stop propose new blocks. 

## ~~Test in localnet~~
1. apply this [patch]
2. start localnet: `./test/debug.sh`
3. check node status via [hmy]:
    - shard0: `hmy blockchain latest-header`
    - shard1: `hmy blockchain latest-header -n http://127.0.0.1:9502`
4. after epoch 4(reduction hardfork in localnet), kill all nodes in shard1 via [bash-script]: `./killShard1.sh`
6. check if shard0 is work fine via [hmy].

## Test in Testnet
1. before shard reduction epoch: if a validator spans multi shard, check and remove its blskey on s2 and s3, otherwise, it will be marked as inactive after reduction-epoch.
2. wait for the reduction-epoch.
3. stop all nodes on s2 and s3.
4. check if s0 and s1 is work fine.

## TODO
- [x] Handle blskey on shard2 and shard3. 
- [x] Handle harmony inner blskey on shard2 and shard3. 
- [x] Test it in local
- [x] test it on testnet

[patch]: https://gist.github.com/peekpi/cdd956433a0756dcb295128c541284d9
[hmy]: https://docs.harmony.one/home/network/validators/node-setup/hmy-cli-download
[bash-script]: https://gist.github.com/peekpi/cdd956433a0756dcb295128c541284d9#kill-shard1-nodes
[createvalidator]: https://gist.github.com/peekpi/cdd956433a0756dcb295128c541284d9#file-createvalidator-sh
